src/metis/http/gist.lua

```diff
@@ -0,0 +1,285 @@
+        if not handle then error("Could not connect to api.github.com.") end
+        if handle.getResponseCode() ~= 200 then
+            handle.close()
+            error("Failed to download file data.")
+        end
+        local d = handle.readAll()
+        handle.close()
+        return d
+    end
+end
+
+local function setTextColor(c) if term.isColor() then term.setTextColor(c) elseif c == colors.white or c == colors.yellow then term.setTextColor(colors.white) else term.setTextColor(colors.lightGray) end end
+
+local function requestAuth(headers, interactive)
+    if settings.get("gist.id") ~= nil then
+        headers.Authorization = "token " .. settings.get("gist.id")
+        return true
+    elseif interactive then
+        setTextColor(colors.yellow)
+        write("You need to add a Personal Access Token (PAK) to upload Gists. Follow the instructions at ")
+        setTextColor(colors.blue)
+        write("https://tinyurl.com/GitHubPAK")
+        setTextColor(colors.yellow)
+        write(" to generate one. Make sure to check the '")
+        setTextColor(colors.blue)
+        write("gist")
+        setTextColor(colors.yellow)
+        print("' checkbox on step 7 (under 'Select scopes'). Once done, paste it here.")
+        setTextColor(colors.lime)
+        write("PAK: ")
+        setTextColor(colors.white)
+        local pak = read()
+        if pak == nil or pak == "" then error("Invalid PAK, please try again.") end
+        settings.set("gist.id", pak)
+        settings.save(".settings")
+        headers.Authorization = "token " .. pak
+        return true
+    end
+    return false
+end
+
+-- User API - this can be loaded with require "cc.http.gist"
+
+-- ID can be either just the gist ID or a gist ID followed by a slash and a file name
+-- * If a file name is specified, retrieves that file
+-- * Otherwise, if there's only one file, retrieves that file
+-- * Otherwise, if there's a file named 'init.lua', retrieves 'init.lua'
+-- * Otherwise, if there's more than one file but only one *.lua file, retrieves the Lua file
+-- * Otherwise, retrieves the first Lua file alphabetically (with a warning)
+-- * Otherwise, fails
+
+--- Retrieves one file from a Gist using the specified ID.
+-- @tparam string id The Gist ID to download from. See above comments for more details.
+-- @tparam[opt] function progress A function to use to report status messages.
+-- @treturn string|nil The contents of the specified Gist file, or nil on error.
+-- @treturn string|nil The name of the file that was chosen to be downloaded, or a message on error.
+function gist.get(id, progress)
+    expect(1, id, "string")
+    expect(2, progress, "function", "nil")
+    progress = progress or emptyfn
+    local file
+    if id:find("/") ~= nil then id, file = id:match("^([0-9A-Fa-f:]+)/(.+)$") end
+    if id == nil or not id:match("^[0-9A-Fa-f][0-9A-Fa-f:]+[0-9A-Fa-f]$") then error("bad argument #1 to 'get' (invalid ID)", 2) end
```

Comment:
✨ You can use `%x` rather than `[0-9A-Fa-f]`.

Reply:
I forgot that `%x` existed when I first wrote this :P. The `:` is for revisions - if you specify an ID like `0123456789abcdef:fedcba9876543210` it will fetch from revision `fedcba9876543210` instead of `HEAD`.